File: static/gallery_res/demo27/demo27.py
"""
Ω = [`e₁` `e₂`][`k₁`   0
                 0    `k₂`] [`e₁`ᵀ
			                 `e₂`ᵀ]

where

`k₁`: ℝ  : control the desired kernel variance in either edge or orthogonal direction
`k₂`: ℝ  : control the desired kernel variance in either edge or orthogonal direction
`e₁`: ℝ ^ 3: orthogonal direction vectors
`e₂`: ℝ ^ 3: orthogonal direction vectors
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


def demo27(k₁, k₂, e₁, e₂):
    """
    :param :k₁ : ℝ  : control the desired kernel variance in either edge or orthogonal direction
    :param :k₂ : ℝ  : control the desired kernel variance in either edge or orthogonal direction
    :param :e₁ : ℝ ^ 3: orthogonal direction vectors
    :param :e₂ : ℝ ^ 3: orthogonal direction vectors
    """
    e₁ = np.asarray(e₁, dtype=np.float64)
    e₂ = np.asarray(e₂, dtype=np.float64)

    assert np.ndim(k₁) == 0
    assert np.ndim(k₂) == 0
    assert e₁.shape == (3,)
    assert e₂.shape == (3,)

    _Ω_0 = np.hstack(((e₁).reshape(3, 1), (e₂).reshape(3, 1)))
    _Ω_1 = np.zeros((2, 2))
    _Ω_1[0] = [k₁, 0]
    _Ω_1[1] = [0, k₂]
    _Ω_2 = np.vstack((e₁.T.reshape(1, 3), e₂.T.reshape(1, 3)))
    Ω = _Ω_0 @ _Ω_1 @ _Ω_2

    return Ω


def generateRandomData():
    k₁ = np.random.randn()
    k₂ = np.random.randn()
    e₁ = np.random.randn(3)
    e₂ = np.random.randn(3)
    return k₁, k₂, e₁, e₂


if __name__ == '__main__':
    k₁, k₂, e₁, e₂ = generateRandomData()
    print("k₁:", k₁)
    print("k₂:", k₂)
    print("e₁:", e₁)
    print("e₂:", e₂)
    func_value = demo27(k₁, k₂, e₁, e₂)
    print("func_value: ", func_value)

File: static/gallery_res/demo27/demo27.cpp
/*
Ω = [`e₁` `e₂`][`k₁`   0
                 0    `k₂`] [`e₁`ᵀ
			                 `e₂`ᵀ]

where

`k₁`: ℝ  : control the desired kernel variance in either edge or orthogonal direction
`k₂`: ℝ  : control the desired kernel variance in either edge or orthogonal direction
`e₁`: ℝ ^ 3: orthogonal direction vectors
`e₂`: ℝ ^ 3: orthogonal direction vectors
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

/**
 * demo27
 *
 * @param k₁  ℝ  : control the desired kernel variance in either edge or orthogonal direction
 * @param k₂  ℝ  : control the desired kernel variance in either edge or orthogonal direction
 * @param e₁  ℝ ^ 3: orthogonal direction vectors
 * @param e₂  ℝ ^ 3: orthogonal direction vectors
 * @return Ω
 */
Eigen::Matrix<double, 3, 3> demo27(
    const double & k₁,
    const double & k₂,
    const Eigen::Matrix<double, 3, 1> & e₁,
    const Eigen::Matrix<double, 3, 1> & e₂)
{
    Eigen::Matrix<double, 3, 2> _Ω_0;
    _Ω_0 << e₁, e₂;
    Eigen::Matrix<double, 2, 2> _Ω_1;
    _Ω_1 << k₁, 0,
    0, k₂;
    Eigen::Matrix<double, 2, 3> _Ω_2;
    _Ω_2 << e₁.transpose(),
    e₂.transpose();
    Eigen::Matrix<double, 3, 3> Ω = _Ω_0 * _Ω_1 * _Ω_2;

    return Ω;
}


void generateRandomData(double & k₁,
    double & k₂,
    Eigen::Matrix<double, 3, 1> & e₁,
    Eigen::Matrix<double, 3, 1> & e₂)
{
    k₁ = rand() % 10;
    k₂ = rand() % 10;
    e₁ = Eigen::VectorXd::Random(3);
    e₂ = Eigen::VectorXd::Random(3);
}


int main(int argc, char *argv[])
{
    double k₁;
    double k₂;
    Eigen::Matrix<double, 3, 1> e₁;
    Eigen::Matrix<double, 3, 1> e₂;
    generateRandomData(k₁, k₂, e₁, e₂);
    Eigen::Matrix<double, 3, 3> func_value = demo27(k₁, k₂, e₁, e₂);
    std::cout<<"func_value:\n"<<func_value<<std::endl;
    return 0;
}